src/zmalloc.c

```diff
@@ -87,10 +88,23 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define update_zmalloc_stat_alloc(__n) used_memory_thread[thread_index] += (__n)
```

Comment:
thread_index is thread local, but used_memory_thread[thread_index] is not.  Without some sort of locking, a thread might get strange value from used_memory_thread[thread_index]  if another thread is updating it

Reply:
@lipzhu I meant corrupted value due to race condition among threads. I did not see (maybe I have missed) how we guarantee reader read either old or new value. The reader can see corrupted value of used_memory_thread[i] if the writer is updating used_memory_thread[i] at the same time

Reply:
Yes, the semantics is a little different from the origin implementation, but do we really need such accurate `used_memory`? The cost is to add the expensive `lock add/sub` instruction in the low-level API like zmalloc/zfree functions, especially most calls to these functions come from the main thread. 
And the `new_value` will be seen in next read.

Reply:
@lipzhu I think data corruption caused by race condition is not about accuracy, the corrupted data can be random from 0 to 2^64-1. This can lead to system correctness or stability issue. 

The original implementation uses atomic number, which leverages support from CPU and is quite efficient as counter that gets updated frequently from different threads.